### PR TITLE
[JIT] Add BatchTensor class

### PR DESCRIPTION
**Pull request by ChunliF**
Add BatchTensor class
- construct from data, mask, dims or construct from list of tensors
- can return a list of tensors from an BatchTensor class

next step: do IR level transformation and operators

### CODE REVIEW

File: torch/csrc/jit/batched/BatchTensor.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+  if(data.dim() != mask.dim() || mask.dim() != int64_t(dims.size()) + 1){
+    throw std::runtime_error("malformed MaskedBatch with data: "
+      + std::string(data.toString()) + " mask: " + mask.toString());
+  }
+  this->data = data;
+  this->mask = mask;
+  this->dims = dims;
+}
+
+BatchTensor::BatchTensor(std::vector<at::Tensor> datalist, std::vector<bool> dims) {
```
Comment (jamesr66a) [inline torch/csrc/jit/batched/BatchTensor.cpp]: Take `datalist` here as const reference.

Later, since you're making an internal copy of `dims`, it's okay to take that as value and move out of this instance into the class member

```
 BatchTensor::BatchTensor(const std::vector<at::Tensor>& datalist, std::vector<bool> dims) { 
```

File: torch/csrc/jit/batched/BatchTensor.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    for(std::size_t j = 0; j < dims.size(); j++){
+      if(dims[j]){
+        data_item = data_item.narrow(j + 1, 0, datalist[i].size(j + 1));
+        mask_item = mask_item.narrow(j + 1, 0, datalist[i].size(j + 1));
+      }
+    }
+    data_item += datalist[i];
+    mask_item.fill_(1);
+  }
+  this->dims = dims;
```
Comment (jamesr66a) [inline torch/csrc/jit/batched/BatchTensor.cpp]:
```
this->dims = std::move(dims);
```

File: torch/csrc/jit/batched/BatchTensor.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+namespace torch { namespace jit {
+
+BatchTensor::BatchTensor(at::Tensor data, at::Tensor mask, std::vector<bool> dims){
+  if(data.dim() != mask.dim() || mask.dim() != int64_t(dims.size()) + 1){
+    throw std::runtime_error("malformed MaskedBatch with data: "
+      + std::string(data.toString()) + " mask: " + mask.toString());
+  }
+  this->data = data;
+  this->mask = mask;
+  this->dims = dims;
```
Comment (jamesr66a) [inline torch/csrc/jit/batched/BatchTensor.cpp]: Since you take `dims` by value, move out of the copy. `this->dims = std::move(dims);`

File: torch/csrc/jit/batched/BatchTensor.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        mask_item = mask_item.narrow(j + 1, 0, datalist[i].size(j + 1));
+      }
+    }
+    data_item += datalist[i];
+    mask_item.fill_(1);
+  }
+  this->dims = dims;
+}
+
+std::vector<at::Tensor> BatchTensor::examples() {
```
Comment (jamesr66a) [inline torch/csrc/jit/batched/BatchTensor.cpp]: Can you add a quick comment explaining what exactly this function is doing?

In particular what `mask_sum` does

**Review by jamesr66a — APPROVED** [review]
Looks fine, I'd just like to see more comments describing what's going on

File: torch/csrc/jit/batched/BatchTensor.cpp (hunk 1 above)
Comment (ChunliF) [reply torch/csrc/jit/batched/BatchTensor.cpp]: As we just discussed, we need `dims` to be Tensor instead of std::vector<bool>. I'll change that to Tensor.

File: torch/csrc/jit/batched/BatchTensor.cpp [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  return result;
+}
+
+void initBatchTensorBindings(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+  py::class_<BatchTensor>(m, "BatchTensor")
+      .def(py::init<at::Tensor, at::Tensor, at::Tensor>())
+      .def(py::init<std::vector<at::Tensor>, at::Tensor>())
+      .def("examples", &BatchTensor::examples)
+      .def("get_data", &BatchTensor::getData)
```
Comment (zdevito) [inline torch/csrc/jit/batched/BatchTensor.cpp]: style nit: in PyTorch we name C++ functions the way they appear in python. So we would have `BatchTensor::get_data`. This violates C++ style, but it is more important to be consistent across C++/Python than it is to follow C++ style.

File: test/test_jit.py [hunk 6]
```diff
@@ -1079,6 +1079,29 @@ def test_fn(ten, mask):
...
+    def assertEqual(self, xs, ybs, prec=None, message='', allow_inf=False):
+        if isinstance(ybs, torch.BatchTensor):
+            self.assertEqual(xs, ybs.examples(), prec, message, allow_inf)
+        elif isinstance(ybs, (list, tuple)):
+            for x, yb in zip(xs, ybs):
+                super(TestBatched, self).assertEqual(x, yb, prec, message, allow_inf)
+        else:
+            super(TestBatched, self).assertEqual(x, y, prec, message, allow_inf)
+
+    def mb_rand(self, *dims):
```
Comment (zdevito) [inline test/test_jit.py]: can you document this? I can't figure out what it is doing.

File: test/test_jit.py [hunk 7]
```diff
@@ -1079,6 +1079,29 @@ def test_fn(ten, mask):
         self.assertEqual(test_fn(ten, mask), traced_test_fn(ten, mask))
 
 
+class TestBatched(TestCase):
+    def assertEqual(self, xs, ybs, prec=None, message='', allow_inf=False):
```
Comment (zdevito) [inline test/test_jit.py]: It think it would be easier to understand these tests if assertEqual didn't attempt to expand a batched tensor automatically.  Then we can use the standard assertEquals:

```
self.assertEqual(xs, batch.examples())
```

File: torch/csrc/jit/batched/BatchTensor.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  }
+  at::Tensor getMask(){
+    return mask;
+  }
+  at::Tensor getDims(){
+    return dims;
+  }
+
+public:
+  at::Tensor data;
```
Comment (zdevito) [inline torch/csrc/jit/batched/BatchTensor.h]: We need documentation here for what data, mask, and dims mean.

**Review by zdevito — COMMENTED** [review]
Code looks good except for a few nits. We need some documentation of the storage formula in the header file though, otherwise it is really hard to follow the code.

**Comment by ezyang** [comment]
@zdevito Just want to confirm that we agreed that we won't try to make BatchTensor wind it's way through all of the existing machinery?

**Comment by zdevito** [comment]
@ezyang -- yes we are going to make it a separate type in python, and represent it as a tuple in the script. In the future when we can define new types more easily we can make it more integrated but it is far to complicated to do it correct right now.

**Review by zdevito — APPROVED** [review]

**Comment by SsnL** [comment]
Sorry to send another email to everyone's inbox. Are we decided that the Python `BatchTensor` lives in top-level `torch.` namespace (vs. something like `torch.jit.BatchTensor`)?

**Comment by apaszke** [comment]
@SsnL good point, it most definitely should be hidden.

**Comment by ChunliF** [comment]
@SsnL  Thanks! Fixed in #9198
